trestle/core/models/elements.py

```diff
@@ -0,0 +1,276 @@
+from trestle.core.err import TrestleError
+
+import yaml
+
+
+class ElementPath:
+    """Element path wrapper of an element.
+
+    This only allows a single wildcard '*' at the end to denote elements of an array of dict
+    """
+
+    PATH_SEPARATOR: str = '.'
+
+    WILDCARD: str = '*'
+
+    def __init__(self, element_path: str):
+        """Initialize an element wrapper."""
+        self._path: List[str] = self._parse(element_path)
+
+        # Initialize variables for lazy processing and caching
+        # This will be processed and cached
+        self._element_name = None
+        self._parent_element_path = None
+
+    def _parse(self, element_path) -> List[str]:
+        """Parse the element path and validate."""
+        parts: List[str] = element_path.split(self.PATH_SEPARATOR)
+
+        for i, part in enumerate(parts):
+            if part == '':
+                raise TrestleError(
+                    f'Invalid path "{element_path}" because having empty path parts between "{self.PATH_SEPARATOR}" \
+                        or in the beginning'
+                )
+            elif part == self.WILDCARD and i != len(parts) - 1:
+                raise TrestleError(f'Invalid path. Wildcard "{self.WILDCARD}" can only be at the end')
+
+        if parts[-1] == self.WILDCARD and len(parts) == 1:
+            raise TrestleError(f'Invalid path {element_path}')
+
+        return parts
+
+    def get(self) -> List[str]:
+        """Return the path components as a list."""
+        return self._path
+
+    def get_first(self) -> str:
+        """Return the first part of the path."""
+        return self._path[0]
+
+    def get_last(self) -> str:
+        """Return the last part of the path."""
+        return self._path[-1]
+
+    def get_element_name(self):
+        """Return the element name from the path."""
+        # if it is available then return otherwise compute
+        if self._element_name is None:
+            element_name = self.get_last()
+            if element_name == self.WILDCARD:
+                element_name = self._path[-2]
+
+            self._element_name = element_name
+
+        return self._element_name
+
+    def get_parent_path(self):
+        """Return the path to the parent element."""
+        # if it is available then return otherwise compute
+        if self._parent_element_path is None:
+            if len(self._path) > 1:
+                parent_path_parts = self._path[:-1]
+                self._parent_element_path = ElementPath(self.PATH_SEPARATOR.join(parent_path_parts))
+
+        return self._parent_element_path
+
+    def __str__(self):
+        """Return string representation of element path."""
+        return self.PATH_SEPARATOR.join(self._path)
+
+    def __eq__(self, other):
+        """Override equality method."""
+        if not isinstance(other, ElementPath):
+            return False
+
+        return self._path == other.get()
+
+
+class Element:
```

Comment:
The biggest question I have here is whether we want to encourage the use of string based specifies in an xpath style model for getting models / elements of whether we should introduce this directly into the OSCAL basemodel. 

The reason why I ask this is here we have introduced a lot of duplication with pydantic. I think it would make more sense to collapse the sense of element and OscalBaseModel and maximise our reliance on pydantic. 

The reason why Is while we have a need for string based accessors when we do we potentially lose typing information. 

This is just a thought.